### DISCUSSION TITLE
No success in building SU2 for windows 10

### DISCUSSION THREAD

**Discussion by ndkazu** [su2code/SU2#1323] [Q&A]
Hello, 
I need to build SU2 on windows for the adjoint capabilities.
I have the exact same problem as the one described in this issue: https://github.com/su2code/SU2/issues/1093
I followed build instructions for windows, used python 3.7 instead of 3.9 to be sure,
I generated a "msmpi.a" file by following the instruction given in the link: https://abhilashreddy.com/writing/3/mpi_instructions.html
added the file to my path, and even manually added it in the \Microsoft SDKs\\MPI\\Lib\\x64\\ folder, nothing is changed.
As you can see in the attached textfile, mpiexec seems to work fine. 
[mpi_lib_problem.txt](https://github.com/su2code/SU2/files/6782110/mpi_lib_problem.txt)

Thanks in advance for your help.
Kazunobu

_Originally posted by @ndkazu in https://github.com/su2code/SU2/issues/1093#issuecomment-876174111_

**Comment by pcarruscag** [comment]
Try giving -Dcustom-mpi=true to meson.py

**Reply by ndkazu** [reply]
Hello Pedro,

I tried and meson stopped complaining.
Now ninja is causing problems, asking for a mpi.h file.
I realised that this file is located in a "Include" folder of Microsoft  SDK, and added it to the path, but the file still cannot be found.
Also, if I manage to build this I really would like to document the process from beginning to end, as I could not find a single complete & working description for SU2 building on windows...
Thanks in advance for your help.

Kazunobu
[ninja_problem.txt](https://github.com/su2code/SU2/files/6784993/ninja_problem.txt)

**Comment by ndkazu** [comment]
Update:
I tried copying the mpi.h file directly into the folders mentioned in the error messages (see the file ninja_problem.txt in my previous post).
ninja does its thing, gives some warnings, and stops/fails the build without any obvious error message. just: build stopped: subcommand failed
[ninja_log.txt](https://github.com/su2code/SU2/files/6785132/ninja_log.txt)
[meson-log.txt](https://github.com/su2code/SU2/files/6785135/meson-log.txt)
[ninja_problem.txt](https://github.com/su2code/SU2/files/6785138/ninja_problem.txt)

**Comment by ndkazu** [comment]
Hello Pedro, Final update for today
After pasting mpi.h in all required folders(before running meson.py), I executed meson.py and then ninja.
ninja seems to run very well for ~20min and then I get a lot of " undefined reference" errors.
I am sending 2 logs: the full ninja log, and the one containing only the errors

Kazunobu
[ninja_log.txt](https://github.com/su2code/SU2/files/6785551/ninja_log.txt)
[ninja_log0.txt](https://github.com/su2code/SU2/files/6785589/ninja_log0.txt)

**Comment by ndkazu** [comment]
Hello Pedro,
I realized that the issue I am experiencing is  the same as the one mentioned here: https://github.com/su2code/SU2/issues/956
adding the tag "--buildtype=release" did not solve it.
Could you please detailed the way you solved this please ?

**Reply by pcarruscag** [reply]
I did not solved it, I mentioned in other issues / discussion that I know nothing about Windows...
As for your problem, if you were on Linux you would use the env vars CXXFLAGS, LDFLAGS, etc. to specify the include paths that the compiler needs to search and the external libraries that are needed to link the code (this is why you get undefined references).

**Comment by ndkazu** [comment]
Hello Pedro, 
First of all: thank you for your answer.
I am now trying to build SU2 in Windows Subsystem Linux (Ubuntu).
I use the following tags:
python3 meson.py -Denable-autodiff=true -Denable-directdiff=true -Dwith-mpi=enabled -Dcustom-mpi=true -Dwith-omp=true -Denable-pywrapper=true -Denable-openblas=true -Dextra-deps=lapack build

And get the following output:
The Meson build system
Version: 0.54.999
Source dir: /mnt/f/SU2/SU2
Build dir: /mnt/f/SU2/SU2/build
Build type: native build
Project name: SU2
Project version: 7.1.1 "Blackbird"
Using 'CC' from environment with value: '/usr/bin/mpicc'
Using 'CXX' from environment with value: '/usr/bin/mpicxx'
Using 'CC' from environment with value: '/usr/bin/mpicc'
C compiler for the host machine: /usr/bin/mpicc (gcc 7.5.0 "gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0")
C linker for the host machine: /usr/bin/mpicc ld.bfd 2.30
Using 'CXX' from environment with value: '/usr/bin/mpicxx'
C++ compiler for the host machine: /usr/bin/mpicxx (gcc 7.5.0 "g++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0")
C++ linker for the host machine: /usr/bin/mpicxx ld.bfd 2.30
Host machine cpu family: x86_64
Host machine cpu: x86_64
Program python3 found: YES (/usr/bin/python3)
Run-time dependency OpenMP for cpp found: YES 4.5
Message: Boost sources found.
Found pkg-config: /usr/bin/pkg-config (0.29.1)
Run-time dependency openblas found: YES 0.2.20
Run-time dependency lapack found: YES 0.2.20+ds
Program swig found: YES (/usr/bin/swig)
Message: Using mpi4py from /home/kazu/.local/lib/python3.6/site-packages/mpi4py/include
Dependency python found: YES (pkgconfig)
Program install.sh found: YES (/mnt/f/SU2/SU2/SU2_PY/pySU2/install.sh)
Message: -------------------------------------------------------------------------
         |    ___ _   _ ___                                                      |
         |   / __| | | |_  )   Release 7.1.1 "Blackbird"                         |
         |   \__ \ |_| |/ /                                                      |
         |   |___/\___//___|   Meson Configuration Summary                       |
         |                                                                       |
         -------------------------------------------------------------------------

         Option          Value
         ---------------------
         TecIO:          true
         CGNS:           true
         AD (reverse):   true
         AD (forward):   true
         Python Wrapper: true
         Intel-MKL:      false
         OpenBlas:       true
         PaStiX:         false
         Mixed Float:    false

         Please be sure to add the $SU2_HOME and $SU2_RUN environment variables,
         and update your $PATH (and $PYTHONPATH if applicable) with $SU2_RUN

         Based on the input to this configuration, add these lines to your .bashrc file:

         export SU2_RUN=/usr/local/bin
         export SU2_HOME=/mnt/f/SU2/SU2
         export PATH=$PATH:$SU2_RUN
         export PYTHONPATH=$PYTHONPATH:$SU2_RUN

         Use './ninja -C build install' to compile and install SU2

Build targets in project: 20

Looks good, MPI detected MPICH is not installed, only OpenMPI.
I run ninja without problem: no error or warning, then run a tutorial in parallel, and there I have a problem.
I ran the following command: 
/usr/bin/mpirun -n 4 /usr/local/bin/SU2_CFD unsteady_naca0012.cfg
on a 24 cores/48threads computer
the cpu is immediately used at 100%, and the calculation is slower than single-core calculation. Very similar to what is described there: https://www.cfd-online.com/Forums/su2-installation/228528-openmpi-su2-simple-parallel-install-gives-problem-iterations.html
MPI is not working, and the only warning are: 
-WARNING: On 4 MPI ranks the coloring efficiency was less than 0.875 (min value was 0.200955).
         Those ranks will now use a fallback strategy, better performance may be possible
         with a different value of config option EDGE_COLORING_GROUP_SIZE (default 512).
-WARNING: SU2 was not compiled for an AVX-capable architecture.
Any idea  about what I can do to fix that please?

Kazun
[wsl_test.txt](https://github.com/su2code/SU2/files/6806521/wsl_test.txt)
obu

**Reply by ndkazu** [reply]
Hey sorry to insist, but I Can't make it work on Ubuntu also then I'm lost.
I tried running the adjoint optimization tutorial in serial (same build as above), and got the following error in the end:

Traceback (most recent call last):
  File "/usr/local/bin/shape_optimization.py", line 183, in <module>
    main()
  File "/usr/local/bin/shape_optimization.py", line 98, in main
    options.nzones      )
  File "/usr/local/bin/shape_optimization.py", line 159, in shape_optimization
    SU2.opt.SLSQP(project,x0,xb,its,accu)
  File "/usr/local/bin/SU2/opt/scipy_tools.py", line 133, in scipy_slsqp
    epsilon        = eps            )
  File "/home/kazu/.local/lib/python3.6/site-packages/scipy/optimize/slsqp.py", line 206, in fmin_slsqp
    constraints=cons, **opts)
  File "/home/kazu/.local/lib/python3.6/site-packages/scipy/optimize/slsqp.py", line 325, in _minimize_slsqp
    for c in cons['ineq']]))
  File "/home/kazu/.local/lib/python3.6/site-packages/scipy/optimize/slsqp.py", line 325, in <listcomp>
    for c in cons['ineq']]))
  File "/usr/local/bin/SU2/opt/scipy_tools.py", line 457, in con_cieq
    cons = project.con_cieq(x)
  File "/usr/local/bin/SU2/opt/project.py", line 257, in con_cieq
    return self._eval(konfig, func,dvs)
  File "/usr/local/bin/SU2/opt/project.py", line 206, in _eval
    vals = design._eval(func,*args)
  File "/usr/local/bin/SU2/eval/design.py", line 147, in _eval
    vals = eval_func(*inputs)
  File "/usr/local/bin/SU2/eval/design.py", line 520, in con_cieq
    func = su2func(this_con,config,state)
  File "/usr/local/bin/SU2/eval/functions.py", line 92, in function
    aerodynamics( config, state )
  File "/usr/local/bin/SU2/eval/functions.py", line 274, in aerodynamics
    info = su2run.direct(config)
  File "/usr/local/bin/SU2/run/direct.py", line 77, in direct
    SU2_CFD(konfig)
  File "/usr/local/bin/SU2/run/interface.py", line 112, in CFD
    run_command( the_Command )
  File "/usr/local/bin/SU2/run/interface.py", line 270, in run_command
    raise exception(message)
SU2.EvaluationFailure: Path = /home/kazu/SU2_test/DESIGNS/DSN_001/DIRECT/,
Command = /usr/local/bin/SU2_CFD config_CFD.cfg
SU2 process returned error '1'

I don't know if it's related, but I hope this will help to solve the MPI problem.
Kazunobu

**Comment by ndkazu** [comment]
Update,
removing the tag "-Dwith-omp=true" solved my Linux Problem.
Summary of important points for Linux build:
- Try to have only one MPI library installed (OpenMPI or MPICH)
- Define Environment variables for MPI libraries(CC, CXX, etc...)
- Do not use the tag "-Dwith-omp=true" when running meson.py
- Avoid using the python scripts ( parallel_computation.py, shape_optimization.py )

The next update is when I find the way to build it successfully on/for Windows.
Kazu.

**Reply by pcarruscag** [reply]
For the message regarding AVX see the docs https://su2code.github.io/docs_v7/Build-SU2-Linux-MacOS/ advanced part involving -march and -mtune.
Regarding "-Dwith-omp=true", you need to specify the number of threads per rank via the CLI option -t (or --threads) or via the env var OMP_NUM_THREADS (as with literally every OpenMP program).
If you change the compilation defaults and run into problems, ask and wait for answers, please do not leave advice for people based on anecdotical evidence.

**Reply by ndkazu** [reply]
Hello Pedro,
Sorry for the bad assumptions. however, I still have a problem: the python scripts do not  seem to work:
when running shape_optimization with the tutorial file, I end up with:
SU2.EvaluationFailure: Path = /home/kazu/SU2_test/DESIGNS/DSN_001/DIRECT/,
Command = /usr/local/bin/SU2_CFD config_CFD.cfg

But, running " /usr/local/bin/SU2_CFD config_CFD.cfg" works fine.
the path is correct, the command is correct, so why does the script fail???

Kazu

**Reply by ndkazu** [reply]
Hello,

I've been testing with the tutorials found on SU2 website: some work right away, and some don't (divergence problem?).
in particular:
- "Unsteady Shape Optimization" in the "Design Features" category. 
- "Epistemic Uncertainty Quantification of RANS predictions of NACA 0012 airfoil" in the "Compressible Flow" category
- "Unsteady NACA00012" in the "Compressible Flow" category
  --Works in serial
  --Fails to find the "SOLVER" parameter in the config file when using parallel_computation.py
  -- "mpirun -n NP" does the work, but then SU2_SOL fails with the error: "Unable to open SU2 restart file restart_flow_00533.dat"

So the python scripts seem to work as expected, except for the cases mentioned above in my build: did I forget anything? is it simply a case configuration problem? can you run/complete the 3 cases above without any modification?

Kazu
